drivers: led: extend lp50xx driver

man-gc opened the pull request:
This PR is based on previous work from PR #46134 and #54594 and has been validated on a custom board using TI LP5012 LED controller.

man-gc committed:
drivers: led: lp503x: extend driver to all lp50xx devices

Add support for LP5009, LP5012, LP5018 and LP5024 devices which only
differ by the number of LEDs they can control.

Also, update application sample to run on all these new supported
devices.

Based on initial work from:
  - Marek Janus <marek.janus@grinn-global.com>
  - Rico Ganahl <rico.ganahl@bytesatwork.ch>

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc committed:
tests: build_all: led: update for new lp50xx driver

Add nodes for all new devices supported by the updated lp50xx driver to
get these enabled as part of the build test.

Based on initial work from:
  - Marek Janus <marek.janus@grinn-global.com>
  - Rico Ganahl <rico.ganahl@bytesatwork.ch>

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc committed:
boards: arm: faze: update for new lp50xx driver

Update compatible property to align it to the updated lp50xx driver.

Based on initial work from: Marek Janus <marek.janus@grinn-global.com>

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc committed:
drivers: led: lp50xx: add reset

The LP50XX family has a specific register to reset the configuration to
default state from any other state. Use this instead of relying on the
manual configuration of registers during startup.

Based on initial work from: Marek Janus <marek.janus@grinn-global.com>

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc committed:
drivers: led: lp50xx: add enable gpio

Some boards may have connected the enable pin of the chipset to a GPIO.
On these boards, it is necessary to configure and set this GPIO before
using the chipset, otherwise the I2C circuitry is disabled.

Based on initial work from:
  - Marek Janus <marek.janus@grinn-global.com>
  - Rico Ganahl <rico.ganahl@bytesatwork.ch>

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc committed:
samples: led_lp50xx: fix color mapping and channel test

Some boards may have less LED child nodes in the DT than the maximum
number of LEDs supported by the chipset. For these boards, the channel
test must skip the missing LEDs when preparing the color buffer and not
exit in error due to missing info.

Also, some boards may have a color mapping that is different from RGB
and this mapping should be used when providing the buffer to
led_set_color().

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

simonguinot commented:
Hi @man-gc. Thank you very much for taking over the https://github.com/zephyrproject-rtos/zephyr/pull/46134 and https://github.com/zephyrproject-rtos/zephyr/pull/54594 PRs. It is very much appreciated. I don't have much time this week to review it. So that will be for next week. Meanwhile, please fix the failed compliance checks.

simonguinot commented:
@man-gc if there are some commits you took from another PR then you have to keep their original authorship and the author's SOB (Signed-Off-By) as well. In the case you made some modifications, you can append your own SOB (and eventually detail the modifications in the commit description message if that's make sense). If you heavily rewrote an existing commit, then you can take the authorship but you have to mention the original author. Since several people worked on this PR, then it is important to give everyone the credit he deserves. Thanks for taking care :)

man-gc commented:
Hi @simonguinot , thanks for the quick reply. I've read your remarks concerning the authorship and I will try to do my best to update the commits with credits where credit is due. This will be a little difficult since the two preceding PRs introduced the same modifications and, in practice, I really only used the first commit of #46134 as a base with the review comments applied. So, maybe I should mention @m-janus in this first commit (author, SOB ?). The fifth and sixth commit are similar to some other commits but have been rewritten from scratch by me during development. The idea is the same as the original ones, but does this count as a rewrite or should I mention the original author ?

Anyway, I will happily apply any remarks regarding this issue.

man-gc commented:
OK, I've tried to check what was in the original PRs and should be attributed to other authors and what constitutes my own "original" contributions. While I cannot deny that I've based my work upon the previous PRs as a source of inspiration, some parts were completely rewritten using other drivers as inspiration (ads1x1x for example) and even with comparison to linux driver. 

So my question is, what is the best way to give credit to the preceding contributions (authorship, SOB, mention) ?

simonguinot commented:
> So my question is, what is the best way to give credit to the preceding contributions (authorship, SOB, mention) ?

If you slightly modified the original patch, then you should retain its original authorship+SOB and append your own SOB indicating that your are conveying this patch.

If you rewrote significant parts of the original patch, then you can take the authorship. And at the end of the commit description (before your SOB) you can add a mention like:

`Based on <developer_name_email>'s patch: <URL to the patch in the original PR>` 

Or: `Based on initial work from <developer_name_email>: <URL to the patch in the original PR>`

Or more simply: `Based on initial work from <developer_name_email>`

You can also have a look at the Linux history for more examples.

Since there is no rule you don't have to worry about the "form". The important thing is to give credit to the developers who worked on it.

bbilas commented:
> > So my question is, what is the best way to give credit to the preceding contributions (authorship, SOB, mention) ?
> 
> If you slightly modified the original patch, then you should retain its original authorship+SOB and append your own SOB indicating that your are conveying this patch.
> 
> If you rewrote significant parts of the original patch, then you can take the authorship. And at the end of the commit description (before your SOB) you can add a mention like:
> 
> `Based on <developer_name_email>'s patch: <URL to the patch in the original PR>`
> 
> Or: `Based on initial work from <developer_name_email>: <URL to the patch in the original PR>`
> 
> Or more simply: `Based on initial work from <developer_name_email>`
> 
> You can also have a look at the Linux history for more examples.
> 
> Since there is no rule you don't have to worry about the "form". The important thing is to give credit to the developers who worked on it.

It's also worth mentioning of `Co-authored-by:`

https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

man-gc commented:
@simonguinot Any news on this PR ?

simonguinot commented:
> @simonguinot Any news on this PR ?

Hi @man-gc,
The initial plan was to review this PR last week. Unfortunately I didn't find the time to do it. So the new plan is to review the PR this week. Since it is a big work I'll split it in several sessions.

man-gc commented:
@simonguinot Thanks for the reply. FYI, I have been rebasing this branch before each push but I read afterwards I shouldn't. I hope this will not make it harder to review this PR and I will comply to the [Contributor Expectations](https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html) from now on.

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,371 @@
+		.max_curr_opt		= DT_INST_PROP(n, max_curr_opt),	\
+		.brightness_base_reg	= bbase_reg,				\
+		.color_base_reg		= cbase_reg,				\
+		.leds_info		= lp##id##_leds_##n,			\
```
Wow, it is big ! `max_leds`, `num_channels`, `brightness_base_reg` and `color_base_reg` are not configuration information. These values ​​are shared/duplicated between all LEDs. Please let's optimize the memory usage a bit and get them out of the lp50xx_config structure.

man-gc replied:
OK, I will remove it for next push using functions with the chip type instead.

man-gc replied:
To remove `num_channels`, I need to populate the `chan_buf` field in `lp50xx_init()`. Is it OK to use k_malloc() inside the driver for this ?

simonguinot replied:
No you don't need to. Since the number of channels can be deduced from the chip type, then you can use a macro to statically initialize the channel buffer. I think it was done this way in #46134.

man-gc replied:
#46134 did actually use a macro but with many ternary operators (and not all devices were supported). I could use a function like I did for the other informations but it won't work statically for initializing the buffer.
Would it be bad practice to initialize the buffer with the `LP5030_NUM_CHANNELS` value to make it work for all devices ?

simonguinot replied:
If I am correct the channel buffer size would be 49 bytes to work for all devices, while `LP5012` only needs a 13 bytes buffer. So yes it is not good. We have to put memory usage and efficiency first. I don't like complicated macros either, but hopefully you can make them understandable and beautiful.

man-gc replied:
OK, I chose the easiest way out by defining all `LP50xx_NUM_CHANNELS` values. This will work for now while witing for a better solution to the `write_channels()` API.

simonguinot replied:
Good for me.

simonguinot commented on include/zephyr/drivers/led/lp50xx.h:
```diff
@@ -0,0 +1,66 @@
+#define LP5024_MAX_LEDS		8
+#define LP5030_MAX_LEDS		10
+#define LP5036_MAX_LEDS		12
+
```
Please add a comment to explain that LP5012 macros are compatible with LP5009, LP5024 macros with LP5018 and LP3036 macros with LP3030.

man-gc replied:
OK

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,371 @@
+		.dev_data	= {						\
+			.min_brightness = 0U,					\
+			.max_brightness = 100U,					\
+		},								\
```
Again, let's not waste memory to store this information. Please use static definition where you need to check minimum/maximum brightness. It will be just fine.

Also, the plan is to increase the maximum brightness to 255 for the LED API. And then all these checks will be removed.

man-gc replied:
OK.

Good news for the increased brightness, it will be easier to set brightness using a raw `int` instead of a percent

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -221,22 +255,10 @@ static int lp50xx_init(const struct device *dev)
-		err = i2c_write_dt(&config->bus, buf, 3);
-	}
+	/* Reset device */
+	err = lp50xx_reset(dev);
```
It is much better. Thanks.

simonguinot reviewed (changes requested):
Hi @man-gc,

Thanks again for this PR ! You did a really great job.

Please see my comments and requests below. I'll go further in the review with the next version.

simonguinot commented:
Let's add more eyes on this PR.

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,475 @@
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(lp50xx);
```
```suggestion
LOG_MODULE_REGISTER(lp50xx, CONFIG_LED_LOG_LEVEL);
```

man-gc replied:
OK

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,475 @@
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/util.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
```
```suggestion

```

man-gc replied:
OK

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,475 @@
+	if (led < config->num_leds) {
+		return &config->leds_info[led];
+	}
+	return NULL;
```
Missing nl before the return statements.

man-gc replied:
OK

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,475 @@
+}
+
+static int lp50xx_get_info(const struct device *dev, uint32_t led,
+			const struct led_info **info)
```
The indentation seems wrong.

man-gc replied:
OK

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,475 @@
+	const struct lp50xx_config *config = dev->config;
+	int err;
+
+	if (!device_is_ready(config->bus.bus)) {
```
```suggestion
	if (!i2c_is_ready_dt(&config->bus)) {
```

man-gc replied:
OK

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,475 @@
+
+	/* Configure GPIO if present */
+	if (config->gpio_enable.port != NULL) {
+		if (!device_is_ready(config->gpio_enable.port)) {
```
```suggestion
		if (!gpio_is_ready_dt(&config->gpio_enable.)) {
```

man-gc replied:
OK

bbilas reviewed (commented):
@man-gc minor comments.
@simonguinot I'm glad that this driver after the looong time is finally on the way to be merged :)

simonguinot commented:
> @simonguinot I'm glad that this driver after the looong time is finally on the way to be merged :)

Don't tell me about it @bbilas ! :) I'm so relieved and thankful to @man-gc for his great work.

man-gc commented on samples/drivers/led_lp50xx/src/main.c:
```diff
@@ -46,21 +47,47 @@ static uint8_t colors[][3] = {
 				"%02x:%02x:%02x, err=%d", led,
-				colors[idx][0], colors[idx][1],
-				colors[idx][2], err);
+				buf[0], buf[1], buf[2], err);
```
OK, next push will contain a `prepare_color_buffer()` function to share this behavior between `run_led_test()` and `run_channel_buffer()`.

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -330,6 +331,34 @@ static int lp50xx_init(const struct device *dev)
+	case PM_DEVICE_ACTION_TURN_OFF:
+		return lp50xx_hw_enable(dev, false);
+	case PM_DEVICE_ACTION_TURN_ON:
+		err = lp50xx_hw_enable(dev, true);
```
`gpio_enable` is not required, so returning 0 seems the right thing to do when it is not present in the devicetree.

simonguinot replied:
I am not familiar with the PM API but I think there is a misunderstanding here. `TURN_ON` and `TURN_OFF` actions are only triggered by a power domain when it is resumed or suspended. So the device/driver can prepare for the power to be removed/added. That's what not we are doing here.

So eventually we should move `gpio_enable` under the `SUSPEND` / `RESUME` action handling and return 0 for  `TURN_ON` / `TURN_OFF` actions (because there is indeed nothing to do). the  What do you think ?

man-gc replied:
Maybe I can use the same behavior as `lp5569`, i.e. use the same action for `TURN_ON`/`RESUME` and `TURN_OFF`/`SUSPEND`, but configuration will be lost each time if I don't keep a copy of the registers to restore inside `lp50xx_data`, am I right ?

simonguinot replied:
No the implementation of the `lp5569` driver doesn't look good. See the "Device actions x states" diagram on https://docs.zephyrproject.org/latest/services/pm/device.html. The `SUSPEND` action is always executed before the `TURN_OFF` action. So the `lp5569` driver runs the same code twice to reach the `OFF` state... Same goes to reach the `ON` state. So I believe it is correct to ignore the `TURN_ON` and `TURN_OFF` PM action in a LED driver.

With your original implementation, `gpio_enable` is basically useless because it needs a power domain to go off (which is unlikely to happen).

But if you use `gpio_enable` when suspending the device, then you are indeed in trouble to restore the registers when resuming. In addition, suspending a device is actually quite different than powering it off...

There is another thing to consider: some apps may not want the LEDs to go off when the device is suspended. An application could even use the LEDs to notify users about the power status. But unfortunately there is no API to disable PM support for a given device/driver...

That's why in https://github.com/zephyrproject-rtos/zephyr/pull/56673, @danieldegrasse choose to implement blanking support through a private API. See the discussion https://github.com/zephyrproject-rtos/zephyr/pull/56673/commits/99bb95315b84f5a3da8bb0a5017d169667dd66ed#r1250848989. Even if I don't like much the idea, I had nothing better to suggest. Maybe we could do the same here with `gpio_enable` ?

simonguinot replied:
Hi @ceolin and @gmarull. Please can you offer some guidance here ?

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -382,10 +378,10 @@ static const struct led_driver_api lp50xx_led_api = {
 
 #define LP50XX_DEVICE(n, id, nchans, rreg, bbase_reg, cbase_reg)		\
-	DT_INST_FOREACH_CHILD(n, COLOR_MAPPING)					\
+	DT_INST_FOREACH_CHILD_STATUS_OKAY(n, COLOR_MAPPING)			\
```
OK, I will remove it for next push. I've been testing the possibility to disable LEDs in the devicetree but if it won't be used, there is no need to keep this code.

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -330,6 +331,34 @@ static int lp50xx_init(const struct device *dev)
+	case PM_DEVICE_ACTION_SUSPEND:
+		return lp50xx_enable(dev, false);
+	case PM_DEVICE_ACTION_RESUME:
+		return lp50xx_enable(dev, true);
```
I would rather not use fallthrough action in that case. I prefer readability over conciseness to avoid any potential side-effect in the future.

simonguinot replied:
OK

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,371 @@
+#define LP5024_OUT0_COLOR		0x0f
+
+/* LP5030/36 registers */
+#define LP5036_LED_CONFIG1		0x03
```
This register is used for bank management of LEDs. It is the same as `LP50xx_LED_CONFIG0`. I know bank management is not available for now, but should I leave these two definitions outside ?

simonguinot replied:
Since bank management is only available for apps using the channel API then maybe we should move them into `include/zephyr/drivers/led/lp50xx.h`.

man-gc replied:
I think we are not talking about the same thing. When I use the term _bank management_, I am referring to a specific feature of the chipset where multiple LEDs could be driven as one. This would require an update to the devicetree to handle a child node with a `reg` array (like Linux does in the corresponding driver).

This will also have an impact on the channel API because it will change the base address to use and/or it will introduce some gaps when using the channel API (mapping is bank_brightness, bank_red, bank_green, bank_blue, led[0;n]_brightness, led0_red, led0_green, led0_blue, ..., ledn_red, ledn_green, ledn_blue). That is way I kept the feature disabled for now.

simonguinot replied:
We are talking about the same thing :) I am well aware of the bank management feature of the LP503{0,6} controllers. There is actually no support for this feature in the LED API. So the bank management registers are useless in the driver itself.

Before this PR the bank registers were exposed via the `write_channels` API. That's why I said that their definitions belong to `include/zephyr/drivers/led/lp50xx.h`. But I now realize it is no longer the case and it is quite a change... Please explain a bit more why we can't keep that.

man-gc replied:
> We are talking about the same thing :) I am well aware of the bank management feature of the LP503{0,6} controllers.
OK, I just wanted to avoid misunderstandings.
> There is actually no support for this feature in the LED API. So the bank registers definitions are useless in the driver itself.
Using the bank registers should be possible with a `reg` array in the devicetree presented as a single LED through the LED API. This is what the Linux driver is doing and I think that should work in Zephyr too. But I think this feature should be added in another PR since the extension of this driver has waited long enough.
> Before this PR the bank registers were exposed via the `write_channels` API. That's why I said that their definitions belong to `include/zephyr/drivers/led/lp50xx.h`. But I now realize it is no longer the case and it is quite a change... Please explain a bit more why we can't keep that.
I did not understand that the `write_channels()` API allowed to use the bank registers. But since it is a raw access, I get your point. However, I think the problem with this API is that it relies too much on the knowledge of the chip and that will make an application using this driver not really portable. But I may have understood wrongly.

If it is too much a change, I can try to find a way to make it work again. Or find another way to make the bank feature available.

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,371 @@
+#define LP5036_LED_CONFIG1		0x03
+#define LP5036_LED0_BRIGHTNESS		0x08
+#define LP5036_OUT0_COLOR		0x14
+
```
`include/zephyr/drivers/led/lp50xx.h` is present because I have been struggling with the channel API which requires to expose some internal registers to make it work for this component (the base addresses of brightness and color registers). Which is why the info is duplicated and I would like to keep it that way because I have been thinking about a solution that would hide these registers and make the channel API more abstract.

I have been thinking of submitting an RFC (not done yet) to make the channel API more like the sensor interface using an enum as an argument to `led_write_channels()` to indicate which channels should be updated. I know that will make it less like the led_strip API but it should be more obvious for this kind of component (which seems to be the only one to use it ?)

simonguinot replied:
It is OK if you wan to keep this registers definitions duplicated. 

About the channel API, note that the [IS31FL3733 driver](https://github.com/zephyrproject-rtos/zephyr/pull/56673) is now using it.

One of the original intention of the channel API was to give users a direct access to the hardware channel registers. Also it was designed to be simple (at least for the driver) and compatible with the LED strip API. For my part I am not a big fan of the idea of abstracting the channels registers. If you want, I am open to discuss it on discord.

man-gc replied:
OK, so to keep the current design, the `write_channels` API should use the LP50XX_LED_CONFIG0 register as base channel. This will make the bank feature work through raw accesses.
But for my part, I am not a big fan of having an API for raw access to a device as it may introduce deltas between configuration and register content... I think I will take the opportunity to join you on discord to discuss it.

simonguinot replied:
With the current implementation, the `LP50XX_LED_CONFIGx` registers are not exposed through the `write_channels` API. It starts with the `bank brightness` register. Please just keep it this way for this PR. And yes, we can discuss future changes for the `lp50xx` driver and the `write_channel` API.

MeisterBob replied:
It is absolutely pointless to expose the `bank` registers without exposing the `led_config` registers. I would suggest to expose the `led_config` registers or offer a function in `lp50xx.h` that enables access to bank mode. Something like https://github.com/metratec/zephyr/blob/96c19193b00add6d939a4e0dad96179c7d1fcafe/include/zephyr/drivers/led/lp50xx.h#L71

MeisterBob replied:
If you expose `lp50xx_config.chip` (and `enum lp50xx_chips`) you can define all the registers in `lp50xx.h`. The big advantage is, that the app doesn't need to know which flavour the LP50xx actually is.

See https://github.com/metratec/zephyr/blob/96c19193b00add6d939a4e0dad96179c7d1fcafe/include/zephyr/drivers/led/lp50xx.h#L23

With `LP5012_NUM_CHANNELS`, `LP5024_NUM_CHANNELS` and `LP5036_NUM_CHANNELS` it's still not obvious which one to pick for LP5009, LP5018 and LP5030.

simonguinot replied:
> It is absolutely pointless to expose the `bank` registers without exposing the `led_config` registers.

Yes it is. But the `write_channel` API is about giving raw access to channel registers but not to the LED configuration registers. More work is needed in the driver to allow bank configuration.

The point here is to let this PR just be about adding support for more controller models. This way it can finally be merged (after the failures of https://github.com/zephyrproject-rtos/zephyr/pull/46134 and https://github.com/zephyrproject-rtos/zephyr/pull/54594). Adding support for the hardware bank feature can come next with another PR. It will require further discussions about the `write_channels` API.

man-gc commented:
@simonguinot @bbilas Thank you for your kind words. I made the first round of modifications but I will wait for some answers before I push it to my repo to start a new round.

gmarull commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,475 @@
+		return lp50xx_enable(dev, false);
+	case PM_DEVICE_ACTION_RESUME:
+		return lp50xx_enable(dev, true);
+	case PM_DEVICE_ACTION_TURN_OFF:
+		return lp50xx_hw_enable(dev, false);
+	case PM_DEVICE_ACTION_TURN_ON:
+		err = lp50xx_hw_enable(dev, true);
```
this is reserved for power domains

simonguinot replied:
Thanks for your help @gmarull.

I do have a question for you:

On some systems we may need to disable PM support for a LED device/driver. For example, a LED can be used to display a pattern when the device runs in a low-power mode. In your opinion, how could we achieve this ?

man-gc replied:
OK, I've read the documentation through and through and I don't understand why this implementation could not work (except for the backup of registers to restore configuration). Why wouldn't you add a LED controller to a power domain ?

For your question, is there anything like a "software" power domain ? The LED could be added to one and therefore the application can be responsible for turning it on and off. Since I am new to Zephyr, it may be a silly idea but I think it is the kind of thing that could work.

For the original discussion at hand, I will remove the `TURN_ON`/`TURN_OFF` actions because, as you said, it won't work as expected and needs more work.

simonguinot replied:
@man-gc Note that you can add "gpio enable" support to the `SUSPEND` / `RESUME` PM action handling. But you'll need to restore the configuration. As an alternative, you can give "gpio enable" control to apps via a private API (like in the [is31fl3733 driver](https://github.com/zephyrproject-rtos/zephyr/pull/56673/commits/45605a05e4ec8770793d1ab1fd7ae8b8b036056a).

man-gc committed:
drivers: led: lp50xx: add power management

Enable device power management using the low-power modes of the LP50XX
family.

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc committed:
drivers: led: lp50xx: fix led index

The led identifer should refer to devicetree ordering, not to the index
used by the controller. This way, it will be possible to deactivate
some leds or to reorganize the indexing if necessary.

Signed-off-by: Mathieu Anquetin <mathieu.anquetin@groupe-cahors.com>

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,475 @@
+	struct lp50xx_data *data = dev->data;
+	uint8_t end_channel, max_channels;
+
+	end_channel = start_channel + num_channels;
```
OK, I saw the `end_channel` and fixed it but thanks for the tip.
For the first point, it should not be a problem because I compare the value with `max_channels` which is also zero-based so should be `- 1`

MeisterBob replied:
Yes, works as intended and not a showstopper: it is only confusing when a variable contains a wrong value.

man-gc commented:
OK, I think I have addressed the majority of the reviews and it should be about time for a new round. Let's force push...

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,518 @@
+		return -EINVAL;
+	}
+
+	return -EINVAL;
```
That is superfluous.

man-gc replied:
fixed

bbilas replied:
IMO. it would be better if you removed that line and restore the `default` within the switch case.

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,518 @@
+		return -EINVAL;
+	}
+
+	return -EINVAL;
```
ditto

man-gc replied:
fixed

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,518 @@
+		return -EINVAL;
+	}
+
+	return -EINVAL;
```
ditto

man-gc replied:
fixed

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,518 @@
+		return -EINVAL;
+	}
+
+	return -EINVAL;
```
ditto

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,518 @@
+		return -EINVAL;
+	}
+
+	return -EINVAL;
```
ditto

man-gc replied:
fixed

bbilas commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,518 @@
+		return -ENODEV;
+	}
+
+	if (!IN_RANGE(value,
```
Won't it fit on 1 line?

man-gc replied:
Yes, it fits. Fixed

bbilas reviewed (dismissed):
@man-gc @simonguinot now it's time for the champagne.

simonguinot reviewed (changes requested):
Let's avoid this PR from being merged. I'll review it tomorrow.

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,408 @@
+	struct lp50xx_data *data = dev->data;
+	uint8_t base_channel, end_channel, max_channels;
+
+	base_channel = lp50xx_get_brightness_base_addr(config->chip);
```
NAK. Before this patch the base channel register of the `write_channels` API is "bank brightness". Now it becomes "LED0 brightness". If there are apps using this driver out there, then this change breaks them silently (without even giving an explanation or a warning it in the commit message).

So please remove this change from this patch and even better from this PR. Let's make this PR just be about adding support for more controller models. If needed, you'll can change `lp50xx_write_channels` with another PR.

man-gc replied:
OK, let's not break things

man-gc replied:
After thought, that is why I find the `write_channels()` API difficult to work with for this controller. Since it is giving raw register access, it is necessary to expose it in some way. Since the base addresses are not the same for all chips, things are gonna break or we need to find a way to expose a common macro to use at runtime and rely on the chip id (and ternary operators).

Even if I don't like it, I think I will expose these macros in `include/zephyr/drivers/led/lp50xx.h` for now just to merge this PR and move on to simplify it in the future.

simonguinot replied:
The base address is different but it is handled by the driver. For the application, channels 0, 1, 2 and 3 are "bank brightness", "bank color A", "bank color B" and "bank color C" registers. Is is true for all models, isn't it ? The driver translates the channel index into a register address. At least it is the current implementation. Can't we do the same with more controller models ?

After the LED brightness and LED color channels indexes can be computed using the maximum number of LEDs (which depends on the chip type).

man-gc replied:
> After the LED brightness and LED color channels indexes can be computed using the maximum number of LEDs (which depends on the chip type).

Yes and no, but I am working on it using an extra offset information.
For example, LP5018 and LP5024 share the same register addresses but LP5018 has a maximum of 6 leds and LP5024 has 8. So we can't directly use the chip type and the maximum number of leds to calculate an offset.

simonguinot commented on include/zephyr/drivers/led/lp50xx.h:
```diff
@@ -0,0 +1,66 @@
+ */
+
+/* LED brightness channels. */
+#define LP5012_LED_BRIGHT_CHAN_BASE	0x07
```
In the original `lp503x` driver these definitions are not actual register addresses but indexes starting at 0 with the first register exposed: "bank brightness".

```
/* Bank channels. */
#define LP503X_BANK_CHAN_BASE           0
#define LP503X_BANK_BRIGHT_CHAN         LP503X_BANK_CHAN_BASE
```

Again, we don't want this patch to break user apps. Let's save the API changes and discussions for another PR.

man-gc replied:
OK, I just need to also fix the sample to return to the state before

simonguinot reviewed (changes requested):
Thanks for the changes and your patience @man-gc. We are almost there.

See my change requests below.

simonguinot commented:
> @man-gc @simonguinot now it's time for the champagne.

The champagne is in the fridge @bbilas. It's for soon but not yet :)

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,408 @@
+	default:
+		return -EINVAL;
+	}
+}
```
The only information known at build time is the chip id which gets added to the `lp50xx_config` structure for each device present in the devicetree. Then, the associated register will only be known at runtime since we need to get the value of `dev->config->chip`.

I can replace all the functions with macros with ternary operators like that : 
`#define LP50XX_MAX_LEDS(chip)				\`
`	((chip) <= LP5012) ? LP5012_MAX_LEDS :		\`
`	((chip) <= LP5024) ? LP5024_MAX_LEDS :		\`
`	((chip) <= LP5036) ? LP5036_MAX_LEDS :		\`
`	-EINVAL									 `

but won't inlining the function have the same effect ? Just tell me what you think and I can make the change like above for all functions if you find it better.

simonguinot replied:
Yes you are right. Actually your first approach (i.e. storing this information in `struct lp50xx_config`) was good. I asked you to not waste memory and I said:

> These values ​​are shared/duplicated between all LEDs

I don't know what I was thinking and I was wrong. The cost is only a few bytes per chip in `struct lp50xx_config` (and not per LED) and it was way better than the conditions we are adding in the code now. Please roll back to your previous version for this part. I apologize for the inconvenience.

Nevertheless check that all the fields you add are needed. For example I am not sure for `color_base_reg`. If we know the number of channels and the brightness base register then it can be computed easily. Please double check for the other fields too.

man-gc commented:
> > @man-gc @simonguinot now it's time for the champagne.
> 
> The champagne is in the fridge @bbilas. It's for soon but not yet :)

Yes, I hope next week. There's not much change to make.

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,351 @@
+static uint8_t lp50xx_get_brightness_base_reg(const struct lp50xx_config *config)
+{
+	return LP50XX_BASE_REG(config->chip) + 4;
+}
```
Please come back at your first version. See my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/59852#discussion_r1270848979.

Adding register fields (`brightness_base_reg`, `reset_reg`, etc) to `struct lp50xx_config` was the right approach. It was much better than using runtime switches (as in `lp50xx_get_brightness_base_reg()` above).

man-gc replied:
I saw your previous comment but checking if all registers were necessary made me realize I could calculate offsets based on the number of modules and that was the only information required. So I took this middle ground approach.
I now am able to transform these functions to macros which can be resolved at compile time to populate the fields of the `lp50xx_config` structure. Will that be a better approach ?

simonguinot replied:
Yes it will. I think that using the number of "modules" is a great idea. And it is OK to compute offsets at runtime using the register address and number of modules information stored in `struct lp50xx_config`. Let's just avoid the runtime switches on the chip type.

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,341 @@
+	((base) + 4)
+
+#define LP50XX_OUT0_COLOR(base, nmodules)	\
+	(LP50XX_LED0_BRIGHTNESS(base) + (nmodules))
```
Can't we deduce the bank base from `nmodules` ?

```
#define LP50XX_BANK_BASE(nmodules)      (3 + (nmodules - 1) / 8)

#define LP50XX_LED0_BRIGHTNESS(nmodules)        \
        (LP50XX_BANK_BASE(nmodules) + 4)
        
#define LP50XX_OUT0_COLOR(nmodules)             \
        (LP50XX_LED0_BRIGHTNESS(nmodules) + (nmodules))

#define LP50XX_RESET(nmodules)                  \
        (LP50XX_OUT0_COLOR(nmodules) + LP50XX_COLORS_PER_LED * (nmodules))
```

man-gc replied:
Well spotted

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,341 @@
+	uint8_t num_modules;
+	uint8_t max_leds;
+	uint8_t num_leds;
+	uint8_t base_reg;
```
If yes, then we can remove the `base_reg` field.

simonguinot commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,341 @@
+#define LP50XX_MIN_BRIGHTNESS		0U
+#define LP50XX_MAX_BRIGHTNESS		100U
+
+#define LP5012_NUM_MODULES	4
```
I think that a comment would be helpful to explain what a module is.

simonguinot reviewed (commented):
Hi @man-gc,

Thank you for the changes and for your patience.

I only have a few  comments. Please see them below.

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,341 @@
+
+struct lp50xx_config {
+	struct i2c_dt_spec bus;
+	enum lp50xx_chips chip;
```
OK, could be useful in the future maybe, but let's remove it for now.

simonguinot commented on samples/drivers/led_lp50xx/src/main.c:
```diff
@@ -127,12 +172,23 @@ static int run_channel_test(const struct device *lp50xx_dev,
-			col[0] = colors[idx][0];
-			col[1] = colors[idx][1];
-			col[2] = colors[idx][2];
+			err = led_get_info(lp50xx_dev, led, &info);
```
This doesn't work for the FaZe board or any other board with less LED child nodes in DT than the `max_leds`. Indeed the FaZe board only has 6 LEDs. So `led_get_info` returns `-EINVAL` when `led` is 6. And the sample exists on this error.

man-gc replied:
Last commit should fix it (tested on a custom board by disabling some LEDs in the DT).

simonguinot replied:
Yes it does.

simonguinot reviewed (changes requested):
It is almost good for me. There is just a last bug to fix in the sample.

simonguinot commented on samples/drivers/led_lp50xx/src/main.c:
```diff
@@ -202,6 +203,7 @@ static int run_channel_test(const struct device *lp50xx_dev,
 		k_sleep(SLEEP_DELAY);
 
 		/* Turn LEDs on. */
+		memset(buffer, 0, sizeof(buffer));
```
Is this `memset` (and all the others after) needed ?

man-gc replied:
I added them because I had some glitches on my board due to leftover values between color / brightness channel settings. Maybe not all are required but it creates some sort of pattern.

Do you prefer me to keep the strict minimum ?

simonguinot reviewed (commented):
Please squash this commit with eeb11a822de6 "samples: led_lp503x: update for new lp50xx driver"

man-gc commented:
> Please squash this commit with [eeb11a8](https://github.com/zephyrproject-rtos/zephyr/commit/eeb11a822de6aa475600691f80b3b2e25874b534) "samples: led_lp503x: update for new lp50xx driver"

I cannot squash this commit because it depends on the preceding one [76b35bd](https://github.com/zephyrproject-rtos/zephyr/pull/59852/commits/76b35bdce0020583aba6d2be253d70dd2c7ea4c3) "samples: led_lp503x: fix color mapping". I can squash them together if you want.

MeisterBob reviewed (commented):
I think that all commits in this PR shall be squashed - otherwise we break the bisectability of the samples, they won't build with only one of them applied.

man-gc commented on drivers/led/lp50xx.c:
```diff
@@ -0,0 +1,424 @@
+		return err;
+	}
+
+	k_usleep(LP50XX_ENABLE_DELAY_US);
```
According to datasheet, the delay is shorter (3us). I've added it but I am not sure it is required.

man-gc commented on boards/arm/faze/faze.dts:
```diff
@@ -78,7 +78,7 @@
 	/* TI LP5030 LED controller connected to I2C1. */
 	led_controller_0: lp5030@30 {
-		compatible = "ti,lp5030", "ti,lp503x";
+		compatible = "ti,lp5030", "ti,lp50xx";
```
It was from the original code, I have just replaced the `lp503x` with `lp50xx`. But since this compatible has been removed, let's remove it here also.

MeisterBob commented:
Thank you. If I could, I would approve it.

man-gc commented:
> > I think that all commits in this PR shall be squashed - otherwise we break the bisectability of the samples, they won't build with only one of them applied.
> 
> We used this approach on #46134 and it was not good... So here I was ready to give up on `git-bisect` from the start. But looking at it now, I think that we can squash the driver and the sample patches. Other patches can stay separate.

@simonguinot @MeisterBob Please be verbose about what you want me to squash together. I don't want to mess this PR by squashing too many commits together. I understand the issue with bisectability but having a big patch is not something desirable too.
IMHO, the first 4 commits could be squashed together since they depend on each other. The rest is just new features and/or fixes and should not have any impact during `git-bisect`.

simonguinot commented:
> > Please squash this commit with [eeb11a8](https://github.com/zephyrproject-rtos/zephyr/commit/eeb11a822de6aa475600691f80b3b2e25874b534) "samples: led_lp503x: update for new lp50xx driver"
> 
> I cannot squash this commit because it depends on the preceding one [76b35bd](https://github.com/zephyrproject-rtos/zephyr/pull/59852/commits/76b35bdce0020583aba6d2be253d70dd2c7ea4c3) "samples: led_lp503x: fix color mapping". I can squash them together if you want.

Yes, please squash them. In addition, please squash `"samples: led_lp503x: update for new lp50xx driver"` with `"drivers: led: lp503x: extend driver to all lp50xx devices"` too. And I think it should be enough to allow bisection on this series.

simonguinot reviewed (approved):
Thanks for the excellent work @man-gc !

simonguinot commented:
Hi @bbilas. Please review this PR.

bbilas commented:
> > > Please squash this commit with [eeb11a8](https://github.com/zephyrproject-rtos/zephyr/commit/eeb11a822de6aa475600691f80b3b2e25874b534) "samples: led_lp503x: update for new lp50xx driver"
> > 
> > 
> > I cannot squash this commit because it depends on the preceding one [76b35bd](https://github.com/zephyrproject-rtos/zephyr/pull/59852/commits/76b35bdce0020583aba6d2be253d70dd2c7ea4c3) "samples: led_lp503x: fix color mapping". I can squash them together if you want.
> 
> Yes, please squash them. In addition, please squash `"samples: led_lp503x: update for new lp50xx driver"` with `"drivers: led: lp503x: extend driver to all lp50xx devices"` too. And I think it should be enough to allow bisection on this series.

It seems that changes related to the `ti,lp503x` compatible within `tests/drivers/build_all/led/app.overlay` in commits `"samples: led_lp503x: update for new lp50xx driver"` and `"boards: arm: faze: update for new lp50xx driver"` should be also squashed with "drivers: led: lp503x: extend driver to all lp50xx devices"` to keep the git bisect working. They are only removing the obsolete compatible that is not available anymore. Am I right?

simonguinot commented:
The `"samples: led_lp503x: update for new lp50xx driver"` patch has been squashed with `"drivers: led: lp503x: extend driver to all lp50xx devices"`. So the sample is OK with `git-bisect` for this series.

The "boards: arm: faze: update for new lp50xx driver" patch is only a cleanup for the "ti,lp503x" which is not used anymore. So git-bisect is OK with the FaZe board too. And I think it is acceptable to have the clean-up change in a dedicated patch.

So we are let with "tests: build_all: led: update for new lp50xx driver". And it is true that the `build_all` test is disabled for one commit for this driver. But honestly I am comfortable with that. I was ready from the start to give up on the whole git-bisect thing in order to avoid a bloated patch and a lack of history. So I am happy with the current status.

bbilas commented:
> The `"samples: led_lp503x: update for new lp50xx driver"` patch has been squashed with `"drivers: led: lp503x: extend driver to all lp50xx devices"`. So the sample is OK with `git-bisect` for this series.
> 
> The "boards: arm: faze: update for new lp50xx driver" patch is only a cleanup for the "ti,lp503x" which is not used anymore. So git-bisect is OK with the FaZe board too. And I think it is acceptable to have the clean-up change in a dedicated patch.
> 
> So we are let with "tests: build_all: led: update for new lp50xx driver". And it is true that the `build_all` test is disabled for one commit for this driver. But honestly I am comfortable with that. I was ready from the start to give up on the whole git-bisect thing in order to avoid a bloated patch and a lack of history. So I am happy with the current status.

That sounds fair enough but I would squash `drivers: led: lp50xx: fix led index` with ``"drivers: led: lp503x: extend driver to all lp50xx devices"`` since it fixes the previous commit. We don't want to have a patch that fixes another patch in the same PR, it makes no sense in my opinion.

simonguinot commented:
> > The `"samples: led_lp503x: update for new lp50xx driver"` patch has been squashed with `"drivers: led: lp503x: extend driver to all lp50xx devices"`. So the sample is OK with `git-bisect` for this series.
> > The "boards: arm: faze: update for new lp50xx driver" patch is only a cleanup for the "ti,lp503x" which is not used anymore. So git-bisect is OK with the FaZe board too. And I think it is acceptable to have the clean-up change in a dedicated patch.
> > So we are let with "tests: build_all: led: update for new lp50xx driver". And it is true that the `build_all` test is disabled for one commit for this driver. But honestly I am comfortable with that. I was ready from the start to give up on the whole git-bisect thing in order to avoid a bloated patch and a lack of history. So I am happy with the current status.
> 
> That sounds fair enough but I would squash `drivers: led: lp50xx: fix led index` with `"drivers: led: lp503x: extend driver to all lp50xx devices"` since it fixes the previous commit. We don't want to have a patch that fixes another patch in the same PR, it makes no sense in my opinion.

No no it fixes a bug of mine in the original driver :) It is important to have this one in a separate commit with a proper message giving an explanation.

bbilas reviewed (approved):
@simonguinot @man-gc All right, let's open that damn cold champagne ;)
